File: c/6.z-字形变换.c
/*
 * @lc app=leetcode.cn id=6 lang=c
 *
 * [6] Z 字形变换
 */

// @lc code=start
char* convert(char* s, int numRows) {
    int l = strlen(s);
    if(l == 0 || numRows == 1) return s;

    char* ret = (char*)malloc(sizeof(char) * (l + 1));
    ret[l] = '\0';

    int n = 0;
    for(int row = 0 ; row < numRows; row++)
    {
        bool sign = 1;
        int interval1 = 2 * (numRows - row) - 2;
        int interval2 = 2 * (numRows - 1) - interval1;
        // printf("interval1 : %d && %d\n",interval1,interval2);
        for(int j = row; j < l;)
        {
            // printf("%d : %c , and %d\n",j,s[j],interval1&&interval2);
            ret[n++] = s[j];
            if(sign)
            {
                j += (interval1 && interval2) ? interval1 : interval2 + interval1;
            }
            else
            {
                j += (interval1 && interval2) ? interval2 : interval1 + interval2;
            }
            sign = !sign;
        }
    }
    return ret;
}



// @lc code=end

File: rs/6.z-字形变换.rs
/*
 * @lc app=leetcode.cn id=6 lang=rust
 *
 * [6] Z 字形变换
 */

// @lc code=start
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        let s_len = s.len();
        let num_rows: usize = num_rows as usize;
        if s_len == 0 || num_rows == 1 || num_rows >= s_len
        {
            return s;
        }

        let mut ret = String::new();

        let mut idx : usize = 0;
        let s = s.chars().collect::<Vec<char>>();
        for row in 0..num_rows
        {
            let mut j = row;
            let mut sign = true;
            let interval1 = 2 * (num_rows - row) - 2;
            let interval2 = 2 * (num_rows - 1) - interval1;
            while j < s_len
            {
                ret.push(s[j]);
                idx += 1;
                if interval1 == 0 || interval2 == 0
                {
                    j += interval1 + interval2;
                }
                else
                {
                    if sign
                    {
                        j += interval1;
                    }
                    else
                    {
                        j += interval2;
                    }
                }
                sign = !sign;
            }
        }
        ret
    }
}
// @lc code=end

